text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
+    /// 
+    /// *Since signature size depends on the security level, a single signature can spread out to up to 3 transactions.
+    /// `inputs` is an object that contains all unused addresses of a seed with a sufficient balance.*
+    /// 
+    /// TODO: This function relies on some `Sponge` trait, which is not yet defined. The code below lays out
+    ///       how the sponge is thought to be used.
+    pub fn seal<S: Sponge<Output=BundleHash>>(self, sponge: S) -> Result<SealedOutgoingBundleBuilder, OutgoingBundleBuilderError> {
+        unimplemented!()
+
+        let mut current_index = 0;
+
+        for transaction_draft in bundle_builder {
+            transaction_draft.current_index = current_index++;
+            transaction_draft.last_index = bundle_builder.size() - 1;
+        }
+
+        let final_hash = loop {
+            // Use the `calculate_hash` function defined above.
+            let hash = bundle_builder.calculate_hash();
+            // See paragraphs on Normalization and M-Bug below pseudocode
+            if hash.normalize().find('M') {
+                bundle_builder
+                    .first_transaction()
+                    .increment_obsolete_tag();
+            } else {
+                break hash;
+            }
+        };
+
+        for transaction_draft in &mut bundle_builder {
+            transaction_draft.bundle = final_hash;
+        }
+    }
+}
+
+impl SealedOutgoingBundleBuilder {
+    /// Calculate the bundle hash using some sponge.
+    ///
+    /// A bundle hash ties different transactions together. By having this common hash in their `bundle` field, it makes it
+    /// clear that these transactions should be processed as a whole.
+    ///
+    /// The hash of a bundle is derived from the bundle essence of each of its transactions. A bundle essence is a `486` trits
+    /// subset of the transaction fields.
+    ///
+    /// | Name          | Size      |
+    /// | ------------- | --------- |
+    /// | address       | 243 trits |
+    /// | value         | 81 trits  |
+    /// | obsolete_tag  | 81 trits  |
+    /// | timestamp     | 27 trits  |
+    /// | current_index | 27 trits  |
+    /// | last_index    | 27 trits  |
+    ///
+    /// The bundle hash is generated with a sponge by iterating through the bundle, from `0` to `last_index`, absorbing the
+    /// bundle essence of each transaction and eventually squeezing the bundle hash from the sponge.
+    ///
+    /// NOTE: `iri v1.8.1` uses `Kerl` as a sponge.
+    ///
+    /// NOTE: This function is defined on: `SealedOutgoingBundleBuilder`, `OutgoingBundleBuilder`, and `IncomingBundleBuilder`.
+    ///
+    /// TODO: This function relies on some `Sponge` trait, which is not yet defined. The code below lays out
+    ///       how the sponge is thought to be used.
+    pub fn calculate_hash<S: Sponge<Output=BundleHash>>(&self, mut sponge: S) -> S::Output {
```

Comment:
Is it worth thinking of a default implementation for this ? That's quite some duplication atm.

Reply:
I am not totally sure about this, so I am going to leave this until after we have a `Sponge` trait.